File: app/src/main/java/com/cliqz/browser/settings/PrivacySettingsFragment.java
/*
 * Copyright 2014 A.C.R. Development
 */
package com.cliqz.browser.settings;

import android.app.Activity;
import android.content.DialogInterface;
import android.os.Build;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.Preference;
import android.preference.PreferenceCategory;
import android.support.v7.app.AlertDialog;

import com.cliqz.browser.R;
import com.cliqz.browser.telemetry.TelemetryKeys;
import com.cliqz.browser.utils.HistoryCleaner;

import java.util.HashMap;

public class PrivacySettingsFragment extends BaseSettingsFragment {

    private static final String SETTINGS_LOCATION = "location";
    private static final String SETTINGS_ENABLECOOKIES = "allow_cookies";
    private static final String SETTINGS_SAVEPASSWORD = "password";
    private static final String SETTINGS_CLEARHISTORY = "clear_history";
    private static final String SETTINGS_CLEARFAVORITES = "clear_favorites";
    private static final String SETTINGS_CLEAR_DATA_ON_EXIT = "clear_private_data_exit";
    private static final String SETTINGS_RESTORETOPSITES = "restore_top_sites";
    private static final String SETTINGS_AUTO_FORGET = "autoforget";
    private static final String SETTINGS_ATTRACK = "attrack";

    private static final int PREFERENCE_GROUP_BROWSING_INDEX = 0;
    private static final int PREFERENCE_GROUP_CLEAR_DATA_INDEX = 1;

    private Activity mActivity;

    @SuppressWarnings("FieldCanBeLocal")
    private CheckBoxPreference cblocation, cbenablecookies, cbsavepasswords, cbclearexit,
            cbautoforget, cbattrack;

    private long startTime;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        startTime = System.currentTimeMillis();
        mTelemetry.sendSettingsMenuSignal(TelemetryKeys.PRIVACY, TelemetryKeys.MAIN);
        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preference_privacy);
        mActivity = getActivity();
        initPrefs();
    }

    private void initPrefs() {
        cbattrack = (CheckBoxPreference) findPreference(SETTINGS_ATTRACK);
        cblocation = (CheckBoxPreference) findPreference(SETTINGS_LOCATION);
        cbenablecookies = (CheckBoxPreference) findPreference(SETTINGS_ENABLECOOKIES);
        cbsavepasswords = (CheckBoxPreference) findPreference(SETTINGS_SAVEPASSWORD);
        cbclearexit = (CheckBoxPreference) findPreference(SETTINGS_CLEAR_DATA_ON_EXIT);
        cbautoforget = (CheckBoxPreference) findPreference(SETTINGS_AUTO_FORGET);

        cbattrack.setOnPreferenceChangeListener(this);
        cblocation.setOnPreferenceChangeListener(this);
        cbenablecookies.setOnPreferenceChangeListener(this);
        cbsavepasswords.setOnPreferenceChangeListener(this);
        cbclearexit.setOnPreferenceChangeListener(this);
        cbautoforget.setOnPreferenceChangeListener(this);

        cbattrack.setChecked(mPreferenceManager.isAttrackEnabled());
        cblocation.setChecked(mPreferenceManager.getLocationEnabled());
        cbenablecookies.setChecked(mPreferenceManager.getCookiesEnabled());
        cbsavepasswords.setChecked(mPreferenceManager.getSavePasswordsEnabled());
        cbautoforget.setChecked(mPreferenceManager.isAutoForgetEnabled());
        cbclearexit.setChecked(mPreferenceManager.getCloseTabsExit()
                || mPreferenceManager.getClearCacheExit()
                || mPreferenceManager.getClearCookiesExitEnabled()
                || mPreferenceManager.getClearHistoryExitEnabled());

        final Preference prefClearHistory = findPreference(SETTINGS_CLEARHISTORY);
        prefClearHistory.setOnPreferenceClickListener(this);
        final Preference restoreTopSites = findPreference(SETTINGS_RESTORETOPSITES);
        restoreTopSites.setOnPreferenceClickListener(this);
        final Preference prefClearFavorites = findPreference(SETTINGS_CLEARFAVORITES);
        prefClearFavorites.setOnPreferenceClickListener(this);

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            ((PreferenceCategory) getPreferenceScreen().getPreference(PREFERENCE_GROUP_BROWSING_INDEX))
                    .removePreference(cbattrack);
        }
    }

    @Override
    public boolean onPreferenceClick(Preference preference) {
        switch (preference.getKey()) {
            case SETTINGS_CLEARHISTORY:
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.CLEAR_HISTORY, TelemetryKeys.PRIVACY);
                clearHistoryDialog();
                return true;
            case SETTINGS_CLEARFAVORITES:
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.CLEAR_FAVORITES, TelemetryKeys.PRIVACY);
                clearFavoritesDialog();
                return true;
            case SETTINGS_RESTORETOPSITES:
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.RESTORE_TOPSITES, TelemetryKeys.PRIVACY);
                restoreTopSitesDialog();
                return true;
            default:
                return false;
        }
    }

    private void clearHistoryDialog() {
        final DialogInterface.OnClickListener dialogListener = new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                switch (which) {
                    case DialogInterface.BUTTON_POSITIVE:
                        HistoryCleaner
                                .builder()
                                .setContext(getActivity())
                                .setDeleteFavorites(false)
                                .build()
                                .cleanup();
                        break;
                    default:
                        break;
                }
                dialog.dismiss();
            }
        };
        final AlertDialog.Builder builder = new AlertDialog.Builder(mActivity);
        builder.setTitle(getResources().getString(R.string.clear_history));
        builder.setMessage(getResources().getString(R.string.dialog_history))
                .setPositiveButton(getResources().getString(R.string.yes), dialogListener)
                .setNegativeButton(getResources().getString(R.string.no), dialogListener)
                .show();
    }

    private void clearFavoritesDialog() {
        final DialogInterface.OnClickListener dialogListener = new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                switch (which) {
                    case DialogInterface.BUTTON_POSITIVE:
                        HistoryCleaner
                                .builder()
                                .setContext(getActivity())
                                .setDeleteFavorites(true)
                                .build()
                                .cleanup();
                        break;
                    default:
                        break;
                }
                dialog.dismiss();
            }
        };
        final AlertDialog.Builder builder = new AlertDialog.Builder(mActivity);
        builder.setTitle(getResources().getString(R.string.clear_favorites));
        builder.setMessage(getResources().getString(R.string.dialog_favorites))
                .setPositiveButton(getResources().getString(R.string.yes), dialogListener)
                .setNegativeButton(getResources().getString(R.string.no), dialogListener)
                .show();
    }

    private void clearPasswordsDialog() {
        final DialogInterface.OnClickListener dialogListener = new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                switch (which) {
                    case DialogInterface.BUTTON_POSITIVE:
                        passwordDatabase.clearPasswords();
                        break;
                    default:
                        break;
                }
            }
        };
        final AlertDialog.Builder builder = new AlertDialog.Builder(mActivity);
        builder.setTitle(R.string.clear_passwords);
        builder.setMessage(R.string.clear_passwords_message)
                .setPositiveButton(R.string.yes, dialogListener)
                .setNegativeButton(R.string.no, dialogListener)
                .show();
    }

    private void restoreTopSitesDialog() {
        final AlertDialog.Builder builder = new AlertDialog.Builder(mActivity);
        builder.setTitle(getResources().getString(R.string.restore_top_sites));
        builder.setMessage(getResources().getString(R.string.message_restore_top_sites))
                .setPositiveButton(getResources().getString(R.string.action_ok), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mHistoryDatabase.restoreTopSites();
                    }
                })
                .setNegativeButton(getResources().getString(R.string.action_cancel), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                }).show();
    }

    private void clearDataDialog() {
        final HashMap<Integer, Boolean> valueSet = new HashMap<>();
        valueSet.put(0, mPreferenceManager.getCloseTabsExit());
        valueSet.put(1, mPreferenceManager.getClearHistoryExitEnabled());
        valueSet.put(2, mPreferenceManager.getClearCookiesExitEnabled());
        valueSet.put(3, mPreferenceManager.getClearCacheExit());
        final String[] entries = new String[]{
                mActivity.getString(R.string.open_tabs),
                mActivity.getString(R.string.history),
                mActivity.getString(R.string.cookies),
                mActivity.getString(R.string.cache)
        };
        final boolean[] values = new boolean[]{
                mPreferenceManager.getCloseTabsExit(),
                mPreferenceManager.getClearHistoryExitEnabled(),
                mPreferenceManager.getClearCookiesExitEnabled(),
                mPreferenceManager.getClearCacheExit()
        };
        final DialogInterface.OnMultiChoiceClickListener dialogListener
                = new DialogInterface.OnMultiChoiceClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which, boolean isChecked) {
                valueSet.put(which, isChecked);
            }
        };
        final AlertDialog.Builder builder = new AlertDialog.Builder(mActivity);
        builder.setMultiChoiceItems(entries, values, dialogListener)
                .setPositiveButton(R.string.action_set, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mPreferenceManager.setCloseTabsExit(valueSet.get(0));
                        mPreferenceManager.setClearHistoryExitEnabled(valueSet.get(1));
                        mPreferenceManager.setClearCookiesExitEnabled(valueSet.get(2));
                        mPreferenceManager.setClearCacheExit(valueSet.get(3));
                        if (valueSet.get(0) || valueSet.get(1) || valueSet.get(2) || valueSet.get(3)) {
                            cbclearexit.setChecked(true);
                        } else {
                            cbclearexit.setChecked(false);
                        }
                    }
                })
                .setNegativeButton(R.string.action_cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                })
                .show();
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        // switch preferences
        switch (preference.getKey()) {
            case SETTINGS_ATTRACK:
                mPreferenceManager.setAttrackEnabled((Boolean) newValue);
                cbattrack.setChecked((Boolean) newValue);
                return true;
            case SETTINGS_LOCATION:
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.LOCATION_ACCESS, TelemetryKeys.PRIVACY,
                        (!(boolean) newValue));
                mPreferenceManager.setLocationEnabled((Boolean) newValue);
                cblocation.setChecked((Boolean) newValue);
                return true;
            case SETTINGS_ENABLECOOKIES:
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.ENABLE_COOKIES, TelemetryKeys.PRIVACY,
                        (!(boolean) newValue));
                mPreferenceManager.setCookiesEnabled((Boolean) newValue);
                cbenablecookies.setChecked((Boolean) newValue);
                return true;
            case SETTINGS_SAVEPASSWORD:
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.SAVE_PASSWORDS, TelemetryKeys.PRIVACY,
                        (!(boolean) newValue));
                mPreferenceManager.setSavePasswordsEnabled((Boolean) newValue);
                cbsavepasswords.setChecked((Boolean) newValue);
                if (!((Boolean) newValue)) {
                    clearPasswordsDialog();
                }
                return true;
            case SETTINGS_CLEAR_DATA_ON_EXIT:
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.CLEAR_CACHE, TelemetryKeys.PRIVACY,
                        (!(boolean) newValue));
                clearDataDialog();
                return false;
            case SETTINGS_AUTO_FORGET:
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.AUTO_FORGET_TAB, TelemetryKeys.PRIVACY,
                        (!(boolean) newValue));
                mPreferenceManager.setAutoForgetModeEnabled((Boolean) newValue);
                return true;
            default:
                return false;
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mTelemetry.sendSettingsMenuSignal(TelemetryKeys.PRIVACY, System.currentTimeMillis() - startTime);
    }
}

File: app/src/main/java/com/cliqz/browser/settings/PrivacySettingsFragment.kt
/*
 * Copyright 2014 A.C.R. Development
 */
package com.cliqz.browser.settings

import acr.browser.lightning.utils.WebUtils
import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.preference.CheckBoxPreference
import android.preference.Preference
import android.preference.PreferenceCategory
import android.provider.Settings
import androidx.appcompat.app.AlertDialog
import com.cliqz.browser.R
import com.cliqz.browser.telemetry.TelemetryKeys
import com.cliqz.browser.utils.HistoryCleaner

class PrivacySettingsFragment : BaseSettingsFragment() {

    private var mActivity: Activity? = null

    private var cbenablecookies: CheckBoxPreference? = null
    private var cbsavepasswords: CheckBoxPreference? = null
    private var cbclearexit: CheckBoxPreference? = null
    private var cbautoforget: CheckBoxPreference? = null
    private var cbattrack: CheckBoxPreference? = null
    private var cbsendusagedata: CheckBoxPreference? = null

    private var prpermissions: Preference? = null

    private var startTime: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        startTime = System.currentTimeMillis()
        mTelemetry.sendSettingsMenuSignal(TelemetryKeys.PRIVACY, TelemetryKeys.MAIN)
        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preferences_privacy)
        mActivity = activity
        initPrefs()
    }

    private fun initPrefs() {
        cbattrack = findPreference(SETTINGS_ANTITRACK) as CheckBoxPreference?
        prpermissions = findPreference(SETTINGS_APP_SETTINGS)
        cbenablecookies = findPreference(SETTINGS_ENABLE_COOKIES) as CheckBoxPreference?
        cbsavepasswords = findPreference(SETTINGS_SAVE_PASSWORD) as CheckBoxPreference?
        cbclearexit = findPreference(SETTINGS_CLEAR_DATA_ON_EXIT) as CheckBoxPreference?
        cbautoforget = findPreference(SETTINGS_AUTO_FORGET) as CheckBoxPreference?
        cbsendusagedata = findPreference(SETTINGS_SEND_USAGE_DATA) as CheckBoxPreference?

        cbattrack?.onPreferenceChangeListener = this
        prpermissions?.onPreferenceClickListener = this

        cbenablecookies?.onPreferenceChangeListener = this
        cbsavepasswords?.onPreferenceChangeListener = this
        cbclearexit?.onPreferenceChangeListener = this
        cbautoforget?.onPreferenceChangeListener = this
        cbsendusagedata?.onPreferenceChangeListener = this

        mPreferenceManager.apply {
            cbattrack?.isChecked = isAttrackEnabled
            cbenablecookies?.isChecked = cookiesEnabled
            cbsavepasswords?.isChecked = savePasswordsEnabled
            cbautoforget?.isChecked = isAutoForgetEnabled
            cbclearexit?.isChecked = (  closeTabsExit || clearCacheExit ||
                                        clearCookiesExitEnabled || clearHistoryExitEnabled)
        }

        findPreference(CLEAR_PRIVATE_DATA)?.onPreferenceClickListener = this
        findPreference(SETTINGS_RESTORE_TOP_SITES)?.onPreferenceClickListener = this
        findPreference(SETTINGS_CLEAR_FAVORITES)?.onPreferenceClickListener = this

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            (preferenceScreen.getPreference(PREFERENCE_GROUP_BROWSING_INDEX) as PreferenceCategory)
                    .removePreference(cbattrack)
        }
    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        when (preference.key) {
            CLEAR_PRIVATE_DATA -> {
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.CLEAR_HISTORY, TelemetryKeys.PRIVACY)
                clearDataDialog()
                return true
            }
            SETTINGS_CLEAR_FAVORITES -> {
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.CLEAR_FAVORITES, TelemetryKeys.PRIVACY)
                clearFavoritesDialog()
                return true
            }
            SETTINGS_RESTORE_TOP_SITES -> {
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.RESTORE_TOPSITES, TelemetryKeys.PRIVACY)
                restoreTopSitesDialog()
                return true
            }
            SETTINGS_APP_SETTINGS -> {
                val activity = activity
                val pm = activity?.packageManager
                val packageName = activity?.packageName
                val permissionsUri = if (packageName != null)
                    Uri.fromParts("package", packageName, null)
                else
                    null
                val intent = if (permissionsUri != null)
                    Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, permissionsUri)
                else
                    null
                val intentSupported = intent != null && pm != null &&
                        intent.resolveActivity(pm) != null
                if (intentSupported) {
                    startActivity(intent)
                }
                return true
            }
            else -> return false
        }
    }

    private fun clearHistoryDialog() {
        val dialogListener = DialogInterface.OnClickListener { dialog, which ->
            when (which) {
                DialogInterface.BUTTON_POSITIVE -> HistoryCleaner
                        .builder()
                        .setContext(activity)
                        .setDeleteFavorites(false)
                        .build()
                        .cleanup()
                else -> {
                }
            }
            dialog.dismiss()
        }
        val builder = AlertDialog.Builder(mActivity!!)
        builder.setTitle(resources.getString(R.string.clear_history))
        builder.setMessage(resources.getString(R.string.dialog_history))
                .setPositiveButton(resources.getString(R.string.yes), dialogListener)
                .setNegativeButton(resources.getString(R.string.no), dialogListener)
                .show()
    }

    private fun clearFavoritesDialog() {
        val dialogListener = DialogInterface.OnClickListener { dialog, which ->
            when (which) {
                DialogInterface.BUTTON_POSITIVE -> HistoryCleaner
                        .builder()
                        .setContext(activity)
                        .setDeleteFavorites(true)
                        .build()
                        .cleanup()
                else -> {
                }
            }
            dialog.dismiss()
        }
        val builder = AlertDialog.Builder(mActivity!!)
        builder.setTitle(resources.getString(R.string.clear_favorites))
        builder.setMessage(resources.getString(R.string.dialog_favorites))
                .setPositiveButton(resources.getString(R.string.yes), dialogListener)
                .setNegativeButton(resources.getString(R.string.no), dialogListener)
                .show()
    }

    private fun clearPasswordsDialog() {
        val dialogListener = DialogInterface.OnClickListener { _, which ->
            when (which) {
                DialogInterface.BUTTON_POSITIVE -> {
                    passwordDatabase.clearPasswords()
                    passwordDatabase.clearBlackList()
                }
                else -> {
                }
            }
        }
        val builder = AlertDialog.Builder(mActivity!!)
        builder.setTitle(R.string.clear_passwords)
        builder.setMessage(R.string.clear_passwords_message)
                .setPositiveButton(R.string.yes, dialogListener)
                .setNegativeButton(R.string.no, dialogListener)
                .show()
    }

    private fun restoreTopSitesDialog() {
        val builder = AlertDialog.Builder(mActivity!!)
        builder.setTitle(resources.getString(R.string.restore_top_sites))
        builder.setMessage(resources.getString(R.string.message_restore_top_sites))
                .setPositiveButton(resources.getString(R.string.action_ok)) { _, _ -> mHistoryDatabase.restoreTopSites() }
                .setNegativeButton(resources.getString(R.string.action_cancel)) { dialog, _ -> dialog.dismiss() }.show()
    }

    private fun clearDataOnExitDialog() {
        mPreferenceManager.apply {
            val entries = arrayOf(
                    getString(R.string.open_tabs),
                    getString(R.string.history),
                    getString(R.string.cookies),
                    getString(R.string.cache) )
            val values = booleanArrayOf(
                    closeTabsExit,
                    clearHistoryExitEnabled,
                    clearCookiesExitEnabled,
                    clearCacheExit )
            AlertDialog.Builder(activity)
                    .setMultiChoiceItems(entries, values) { _, which, isChecked ->
                        values[which] = isChecked
                    }
                    .setPositiveButton(R.string.action_set) { _, _ ->
                        closeTabsExit = values[0]
                        clearHistoryExitEnabled = values[1]
                        clearCookiesExitEnabled = values[2]
                        clearCacheExit = values[3]
                        cbclearexit?.isChecked = values.reduce { a, v -> a || v }
                    }
                    .setNegativeButton(R.string.action_cancel) { dialog, _ -> dialog.dismiss() }
                    .show()
        }
    }

    private fun clearDataDialog() {
        val entries = arrayOf(mActivity!!.getString(R.string.history), mActivity!!.getString(R.string.cookies), mActivity!!.getString(R.string.cache), mActivity!!.getString(R.string.clear_passwords))
        val values = BooleanArray(entries.size) { false }
        val actions = arrayOf(
                { mHistoryDatabase.clearHistory(false) },
                { WebUtils.clearCookies(activity) },
                { WebUtils.clearCache(activity) },
                { passwordDatabase.clearPasswords(); passwordDatabase.clearBlackList() } )
        AlertDialog.Builder(mActivity!!)
                .setMultiChoiceItems(entries, null) { _, which, isChecked ->
                    values[which] = isChecked
                }
                .setPositiveButton(R.string.action_clear) { _, _ ->
                    values.forEachIndexed { index, value -> if (value) actions[index]() }
                }
                .setNegativeButton(R.string.action_cancel) { dialog, _ -> dialog.dismiss() }
                .show()
    }

    override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
        // switch preferences
        when (preference.key) {
            SETTINGS_ANTITRACK -> {
                mPreferenceManager.isAttrackEnabled = newValue as Boolean
                cbattrack!!.isChecked = newValue
                return true
            }
            SETTINGS_ENABLE_COOKIES -> {
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.ENABLE_COOKIES, TelemetryKeys.PRIVACY,
                        !(newValue as Boolean))
                mPreferenceManager.cookiesEnabled = newValue
                cbenablecookies!!.isChecked = newValue
                return true
            }
            SETTINGS_SAVE_PASSWORD -> {
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.SAVE_PASSWORDS, TelemetryKeys.PRIVACY,
                        !(newValue as Boolean))
                mPreferenceManager.savePasswordsEnabled = newValue
                cbsavepasswords!!.isChecked = newValue
                if (!newValue) {
                    clearPasswordsDialog()
                }
                return true
            }
            SETTINGS_CLEAR_DATA_ON_EXIT -> {
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.CLEAR_CACHE, TelemetryKeys.PRIVACY,
                        !(newValue as Boolean))
                clearDataOnExitDialog()
                return false
            }
            SETTINGS_AUTO_FORGET -> {
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.AUTO_FORGET_TAB, TelemetryKeys.PRIVACY,
                        !(newValue as Boolean))
                mPreferenceManager.setAutoForgetModeEnabled(newValue)
                return true
            }
            SETTINGS_SEND_USAGE_DATA -> {
                mTelemetry.sendSettingsMenuSignal(TelemetryKeys.SEND_USAGE_DATA,
                        TelemetryKeys.PRIVACY, newValue as Boolean)
                mPreferenceManager.setSendUsageData(newValue)
                return true
            }
            else -> return false
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mTelemetry.sendSettingsMenuSignal(TelemetryKeys.PRIVACY, System.currentTimeMillis() - startTime)
    }

    companion object {

        private const val SETTINGS_APP_SETTINGS = "settings"
        private const val SETTINGS_ENABLE_COOKIES = "allow_cookies"
        private const val SETTINGS_SAVE_PASSWORD = "password"
        private const val CLEAR_PRIVATE_DATA = "clear_private_data"
        private const val SETTINGS_CLEAR_FAVORITES = "clear_favorites"
        private const val SETTINGS_CLEAR_DATA_ON_EXIT = "clear_private_data_exit"
        private const val SETTINGS_RESTORE_TOP_SITES = "restore_top_sites"
        private const val SETTINGS_AUTO_FORGET = "autoforget"
        private const val SETTINGS_ANTITRACK = "attrack"
        private const val SETTINGS_SEND_USAGE_DATA = "send_usage_data"
        private const val PREFERENCE_GROUP_BROWSING_INDEX = 0
    }
}
